File: gateway/src/main/kotlin/com/qianlei/gateway/router/TrieTree.kt
package com.qianlei.gateway.router

import java.util.concurrent.ConcurrentHashMap

/**
 * Trie Tree 实现
 * @author qianlei
 */
class TrieTree<T> {
    private val root = TrieTreeNode<T>()

    fun add(path: List<String>, value: T) {
        var node = root
        for (i in path.indices) {
            val subPath = path[i]
            if (subPath == "*") {
                val newNode = TrieTreeNode(value)
                node.child["*"] = newNode
                node = newNode
            } else {
                val child = node.child[subPath]
                if (child == null) {
                    val newNode = TrieTreeNode<T>()
                    node.child[subPath] = newNode
                    node = newNode
                } else {
                    node = child
                }
            }
            if (i == path.size - 1) {
                node.value = value
            }
        }
    }

    fun del(path: List<String>): T? {
        var ret: T? = null
        var node = root
        for (i in path.indices) {
            val subPath = path[i]
            if (subPath == "*") {
                return node.child.remove("*")?.value
            } else {
                val child = node.child[subPath] ?: return null
                node = child
            }
            if (i == path.size - 1) {
                ret = node.value
                node.value = null
            }
        }
        return ret
    }

    fun get(path: List<String>): T? {
        var ret: T? = null
        var node = root
        node.child["*"]?.let { ret = it.value }
        for (i in path.indices) {
            val subPath = path[i]
            node.child["*"]?.let { ret = it.value }
            node = node.child[subPath] ?: return ret
            if (i == path.size - 1 && node.value != null) {
                ret = node.value
            }
        }
        return ret
    }

    class TrieTreeNode<T>(
        var value: T? = null,
        val child: MutableMap<String, TrieTreeNode<T>> = ConcurrentHashMap()
    )
}

File: gateway/router/trietree/trietree.go
package trietree

type Node struct {
	value interface{}
	child map[string]*Node
}

type TrieTree struct {
	root *Node
}

func NewTrieTree() *TrieTree {
	return &TrieTree{root: &Node{child: make(map[string]*Node)}}
}

func (t *TrieTree) Add(path []string, value interface{}) {
	var node = t.root
	if len(path) == 0 {
		node.value = value
		return
	}
	for i := 0; i < len(path); i++ {
		subPath := path[i]
		if subPath == "*" {
			newNode := &Node{value: value, child: make(map[string]*Node)}
			node.child["*"] = newNode
			node = newNode
		} else {
			child := node.child[subPath]
			if child == nil {
				newNode := &Node{child: make(map[string]*Node)}
				node.child[subPath] = newNode
				node = newNode
			} else {
				node = child
			}
		}
		if i == len(path)-1 {
			node.value = value
		}
	}
}

func (t *TrieTree) Del(path []string) interface{} {
	var ret interface{} = nil
	node := t.root
	if len(path) == 0 {
		ret := node.value
		node.value = nil
		return ret
	}
	for i := 0; i < len(path); i++ {
		subPath := path[i]
		if subPath == "*" {
			ans := node.child["*"]
			delete(node.child, "*")
			return ans.value
		} else {
			child := node.child[subPath]
			if child == nil {
				return nil
			}
			node = child
		}
		if i == len(path)-1 {
			ret = node.value
			node.value = nil
		}
	}
	return ret
}

func (t *TrieTree) Get(path []string) interface{} {
	var ret interface{} = nil
	var node = t.root
	if node.child["*"] != nil {
		ret = node.child["*"].value
	}
	if len(path) == 0 && node.value != nil {
		ret = node.value
	}
	for i := 0; i < len(path); i++ {
		subPath := path[i]
		if node.child["*"] != nil {
			ret = node.child["*"].value
		}
		if node.child[subPath] == nil {
			return ret
		} else {
			node = node.child[subPath]
		}
		if i == len(path)-1 && node.value != nil {
			ret = node.value
		}
	}
	return ret
}
